File: analysis/cellsystems/s2.ts
// The S2 cell system: the sphere divided by the six faces of a cube projected
// onto it, each face a quadtree of exactly-nested quadrilateral cells
// addressed by Hilbert-curve tokens.
//
// It is the proof the contract holds nothing geohash-shaped. Identities do not
// refine by appending characters — the children of 47a1cb are 47a1ca4 and its
// siblings. Cells have edges that are straight on neither projection. And an
// address only means something on a ground that declares what its picture is
// of, so this is the system that can refuse a ground.
//
// The spherical math is s2js's — the Go library ported, test vectors and all,
// and the one third-party dependency this lane keeps (issue #5 §9's pinned
// surface). This file only bridges it to world pixels through the ground's
// declared flattening.

import { s2 } from "s2js";

import type { CellID, CellLabel, GroundedCellSystem, LocatedCell, Pole, Ring } from "./contract.ts";
import type { CellSystem } from "./contract.ts";
import type { Coordinate, Extent, Ground } from "./ground.ts";
import { surfaceExtent } from "./ground.ts";
import { palette } from "./tokens.ts";
import { geoMapping, worldSurface } from "../semconv/geometry.ts";
import type { GeoMapping, World } from "../semconv/geometry.ts";

/**
 * The telescope stops at S2 level 10 — cells a few kilometres across on a
 * Mars-sized body, tokens six characters long — which is depth enough for a
 * couple of thousand pins and shallow enough that every level stays a place
 * rather than a coordinate.
 */
const deepestS2Level = 10;

/**
 * Port levels count the root as 0 and the six faces as 1; S2 counts the faces
 * as level 0. The two bookkeepings meet here and nowhere else.
 */
const portLevel = (level: number): number => level + 1;

const levelOfToken = (id: CellID): number => s2.cellid.level(s2.cellid.fromToken(id));

/**
 * The mapping bridge, remembered per world: world pixels are only meaningful
 * through the flattening the ground declares, and reading it is parsing.
 */
const mappings = new WeakMap<World, GeoMapping | null>();

function mappingOf(ground: Ground): GeoMapping | null {
  const world = ground.world;
  const held = mappings.get(world);
  if (held !== undefined) return held;
  const fresh = geoMapping(world);
  mappings.set(world, fresh);
  return fresh;
}

/** Degrees from a point on the unit sphere. */
function degreesOf(point: s2.Point): readonly [number, number] {
  const ll = s2.LatLng.fromPoint(point);
  return [(ll.lat * 180) / Math.PI, (ll.lng * 180) / Math.PI];
}

/**
 * The leaf cell under a coordinate, remembered against the mapping that
 * produced it: leaves are looked up per pin on every filter pass, and a
 * coordinate is a stable object for exactly as long as its pin is.
 */
const leaves = new WeakMap<object, { mapping: GeoMapping; leaf: bigint }>();

function leafOf(ground: Ground, coordinate: Coordinate): bigint {
  const mapping = mappingOf(ground);
  if (!mapping) return 0n;
  const memo = leaves.get(coordinate);
  if (memo && memo.mapping === mapping) return memo.leaf;
  const [lat, lng] = mapping.toLatLng(coordinate[0], -coordinate[1]);
  const leaf = s2.cellid.fromLatLng(s2.LatLng.fromDegrees(lat, lng));
  leaves.set(coordinate, { mapping, leaf });
  return leaf;
}

/**
 * The boundary as world pixels: the four geodesic edges, each walked in small
 * great-circle steps on the sphere itself, then unwrapped so a loop that
 * crossed the antimeridian stays continuous. The adapters wrap or split it,
 * each in its own way.
 */
function ringOf(ground: Ground, mapping: GeoMapping, id: CellID): Coordinate[] {
  const cellID = s2.cellid.fromToken(id);
  const cell = s2.Cell.fromCellID(cellID);
  const level = s2.cellid.level(cellID);
  const steps = Math.max(2, Math.min(16, 2 ** (4 - level) * 2));
  const surface = surfaceExtent(ground);
  const width = surface[2] - surface[0];
  const points: [number, number][] = [];
  for (let edge = 0; edge < 4; edge++) {
    const from = cell.vertex(edge).vector;
    const to = cell.vertex((edge + 1) % 4).vector;
    for (let step = 0; step < steps; step++) {
      const t = step / steps;
      const x = from.x + (to.x - from.x) * t;
      const y = from.y + (to.y - from.y) * t;
      const z = from.z + (to.z - from.z) * t;
      const norm = Math.hypot(x, y, z) || 1;
      const [lat, lng] = degreesOf(new s2.Point(x / norm, y / norm, z / norm));
      const [worldX, worldY] = mapping.toWorld(lat, lng);
      points.push([worldX, -worldY]);
    }
  }
  // Sequential unwrap: a jump of more than half the world is the seam, not
  // the ground.
  for (let at = 1; at < points.length; at++) {
    const current = points[at];
    const previous = points[at - 1];
    if (!current || !previous) continue;
    const delta = current[0] - previous[0];
    if (delta > width / 2) current[0] -= width;
    else if (delta < -width / 2) current[0] += width;
  }
  // The closing point joins the loop in the frame the walk ended in, not the
  // one it began in: a pole cell's loop accumulates a whole world of
  // longitude, and closing back to the original frame would hand the adapters
  // a segment sweeping the entire planet — a phantom edge the boundary draws
  // and the fill fans spokes to.
  const opening = points[0];
  const ending = points[points.length - 1];
  if (!opening || !ending) return points;
  const closing: [number, number] = [opening[0], opening[1]];
  const delta = closing[0] - ending[0];
  if (delta > width / 2) closing[0] -= width;
  else if (delta < -width / 2) closing[0] += width;
  points.push(closing);
  return points;
}

/**
 * The mapping this ground declares, or a refusal. Every geometric method needs
 * it, and `appliesTo` is the question a caller was supposed to ask first.
 */
function requireMapping(ground: Ground): GeoMapping {
  const mapping = mappingOf(ground);
  if (!mapping) {
    throw new Error(
      "s2: this ground declares no invertible flattening — ask appliesTo(ground) first " +
      "(analysis/cellsystems, issue #5 §5.4)",
    );
  }
  return mapping;
}

/** The surface as a rectangle: the root's own geometry, since the root is the ground. */
function surfaceRing(ground: Ground): Coordinate[] {
  const [minimumX, minimumY, maximumX, maximumY] = surfaceExtent(ground);
  return [
    [minimumX, minimumY],
    [minimumX, maximumY],
    [maximumX, maximumY],
    [maximumX, minimumY],
    [minimumX, minimumY],
  ];
}

// The root is `""` in every system (contract.ts), and S2's token vocabulary
// has no spelling for it — `fromToken("")` answers a cell id that is not the
// sphere and not an error. Every method below that could be handed the root
// answers for the whole ground instead of asking s2js a question it cannot
// hear. None of these cases is reachable from a plan (the root is never a plan
// cell) which is why the oracle never had to have an answer; the contract says
// it should, so this one does. See docs/analysis.md, "the root cases".

/** S2, bound to one ground. */
function grounded(ground: Ground): GroundedCellSystem {
  const system: GroundedCellSystem = {
    level(id: CellID): number {
      return id === "" ? 0 : portLevel(levelOfToken(id));
    },

    parent(id: CellID): CellID {
      if (id === "") return "";
      const cellID = s2.cellid.fromToken(id);
      const level = s2.cellid.level(cellID);
      if (level === 0) return "";
      return s2.cellid.toToken(s2.cellid.parent(cellID, level - 1));
    },

    children(id: CellID): CellID[] {
      if (id === "") {
        return [0, 1, 2, 3, 4, 5].map((face) => s2.cellid.toToken(s2.cellid.fromFace(face)));
      }
      return s2.cellid.children(s2.cellid.fromToken(id)).map(s2.cellid.toToken);
    },

    childIndex(id: CellID): number {
      if (id === "") return 0;
      const cellID = s2.cellid.fromToken(id);
      const level = s2.cellid.level(cellID);
      return level === 0 ? s2.cellid.face(cellID) : s2.cellid.childPosition(cellID, level);
    },

    // The root is the whole sphere: it holds every point the ground can name.
    contains(id: CellID, at: Coordinate): boolean {
      const leaf = leafOf(ground, at);
      if (id === "") return leaf !== 0n;
      return s2.cellid.contains(s2.cellid.fromToken(id), leaf);
    },

    descendTarget(id: CellID, at: Coordinate): CellID {
      const leaf = leafOf(ground, at);
      if (leaf === 0n) return "";
      const target = id === "" ? 0 : levelOfToken(id) + 1;
      if (target > deepestS2Level) return "";
      return s2.cellid.toToken(s2.cellid.parent(leaf, target));
    },

    // The root is the whole ground: six faces cover a sphere, and their union
    // in world pixels is the picture itself.
    bbox(id: CellID): Extent {
      if (id === "") return surfaceExtent(ground);
      let minimumX = Infinity;
      let minimumY = Infinity;
      let maximumX = -Infinity;
      let maximumY = -Infinity;
      for (const [x, y] of system.ring(id)) {
        minimumX = Math.min(minimumX, x);
        minimumY = Math.min(minimumY, y);
        maximumX = Math.max(maximumX, x);
        maximumY = Math.max(maximumY, y);
      }
      return [minimumX, minimumY, maximumX, maximumY];
    },

    center(id: CellID): Coordinate {
      const mapping = requireMapping(ground);
      if (id === "") {
        const [minimumX, minimumY, maximumX, maximumY] = surfaceExtent(ground);
        return [(minimumX + maximumX) / 2, (minimumY + maximumY) / 2];
      }
      const [lat, lng] = degreesOf(s2.Cell.fromCellID(s2.cellid.fromToken(id)).center());
      const [x, y] = mapping.toWorld(lat, lng);
      return [x, -y];
    },

    ring(id: CellID): Ring {
      const mapping = requireMapping(ground);
      if (id === "") return surfaceRing(ground);
      return ringOf(ground, mapping, id);
    },

    // The root circles both poles, and no single answer says so; nothing asks,
    // because a pole closure is a question about a cell being drawn.
    poleContained(id: CellID): Pole | null {
      if (id === "") return null;
      const cellID = s2.cellid.fromToken(id);
      if (s2.cellid.contains(cellID, s2.cellid.fromLatLng(s2.LatLng.fromDegrees(90, 0)))) {
        return "north";
      }
      if (s2.cellid.contains(cellID, s2.cellid.fromLatLng(s2.LatLng.fromDegrees(-90, 0)))) {
        return "south";
      }
      return null;
    },

    label(id: CellID): CellLabel {
      return { context: id.slice(0, -1), principal: id.slice(-1) };
    },

    // Sibling tokens can share their final character, so the accent comes from
    // the sibling ordinal instead — neighbours always differ.
    colorKey(id: CellID): number {
      return system.childIndex(id) % palette.length;
    },

    normalizeInput(text: string): string {
      return [...text.toLocaleLowerCase()]
        .filter((character) => "0123456789abcdef".includes(character))
        .slice(0, 6)
        .join("");
    },

    // A partial token is not yet a place: the field holds the draft and the
    // session stays put until the address parses whole. A deeper token clamps
    // to the telescope's floor, re-tokenized so the spelling is canonical.
    parseInput(text: string): CellID | null {
      if (text === "") return "";
      const cellID = s2.cellid.fromToken(text);
      if (cellID === 0n || !s2.cellid.valid(cellID)) return null;
      const level = s2.cellid.level(cellID);
      if (level > deepestS2Level) {
        return s2.cellid.toToken(s2.cellid.parent(cellID, deepestS2Level));
      }
      return s2.cellid.toToken(cellID);
    },

    locate(at: Coordinate): LocatedCell | null {
      const leaf = leafOf(ground, at);
      if (leaf === 0n) return null;
      return { label: "S2", value: s2.cellid.toToken(s2.cellid.parent(leaf, deepestS2Level)) };
    },
  };
  return system;
}

export const s2System: CellSystem = {
  slug: "s2",
  name: "S2",
  short: "S2",

  // Only a ground that declares what its picture is of — a sphere, through any
  // invertible flattening — can be divided geodesically.
  appliesTo(ground: Ground): boolean {
    return worldSurface(ground.world) === "sphere" && geoMapping(ground.world) !== null;
  },

  maxLevel(): number {
    return portLevel(deepestS2Level);
  },

  // A level-10 token is six characters; the field takes no more.
  inputLength(): number {
    return 6;
  },

  on: grounded,
};

File: internal/app/cells/s2.go
package cells

import (
	"strings"

	"github.com/golang/geo/s2"

	"github.com/FelineStateMachine/atlas/format/semconv"
)

// The S2 side: the sphere divided by the six faces of a cube projected onto
// it, each face a quadtree of exactly-nested quadrilaterals addressed by
// Hilbert-curve tokens.
//
// Nothing spherical is written here. `github.com/golang/geo/s2` is the
// canonical Google library and the seam's s2js is a port of it, so the two
// lanes hold the same arithmetic by provenance rather than by agreement —
// which is the whole reason the server may answer a containment question at
// all instead of asking the seam for its answer. What this file adds is the
// two things the port has that the library does not: the bridge from world
// pixels to the sphere through the ground's declared flattening, and the level
// bookkeeping in which the root is a place.
//
// LEVELS. The port counts the root as 0 and the six faces as 1; S2 counts the
// faces as 0. The two bookkeepings meet in [S2Level] and nowhere else, exactly
// as they meet in `portLevel` in analysis/cellsystems/s2.ts.
//
// THE ROOT. The root is `""` in every system, and S2's token vocabulary has no
// spelling for it — `CellIDFromToken("")` answers a cell id that is neither
// the sphere nor an error. Every function below that could be handed the root
// answers for the whole ground instead of asking the library a question it
// cannot hear.

// deepestS2Level is where the telescope stops: cells a few kilometres across
// on a Mars-sized body, tokens six characters long. It is the seam's floor
// (analysis/cellsystems/s2.ts) and is spelled here so that an address arriving
// from a request is held to the same floor the navigator would have produced.
const deepestS2Level = 10

// Mapping is a world's declared flattening, read once and inverted per point.
//
// A geodesic system addresses a sphere, and a packed position is a pixel on a
// raster. The mapping is the whole of what turns one into the other, and a
// world that declares none cannot be divided by S2 at all — which is the
// refusal [Applicable] makes on the session's behalf, before any of this is
// reached.
type Mapping struct{ window semconv.Equirect }

// MappingOf reads the world's declared flattening out of its attributes, or
// refuses. It mirrors `geoMapping` in analysis/semconv/geometry.ts, including
// its one subtlety: the window is only read when `atlas.geometry.projection`
// says which flattening the numbers are for. A world carrying an equirect
// window under some other projection is declaring something this cannot read,
// and a plausible answer would be worse than none.
func MappingOf(attrs map[string]string) (Mapping, bool) {
	if attrs[semconv.KeyGeometryProjection] != semconv.ProjectionEquirect {
		return Mapping{}, false
	}
	window, declared, err := semconv.EquirectOf(attrs)
	if !declared || err != nil {
		return Mapping{}, false
	}
	return Mapping{window: window}, true
}

// LatLng puts a point back on the sphere: world pixels in, true degrees out.
//
// The x, y are the application's own — y **negative-down**, the space a pin's
// position lives in — and the sign is flipped here, on the one line where the
// two conventions meet. [semconv.Equirect.Invert] reads y down from the top of
// the raster window, which is the convention the attributes are written in and
// the one the seam's mapping speaks (see the same flip at `leafOf` in
// analysis/cellsystems/s2.ts).
func (m Mapping) LatLng(x, y float64) (lat, lon float64) {
	return m.window.Invert(x, -y)
}

// S2Held is the held-cell question for an S2 address: the point goes back on
// the sphere, becomes the leaf cell holding it, and the held cell is asked
// whether that leaf is one of its descendants.
//
// Two addresses hold nothing rather than something. An empty token is the
// root, and the root is the whole sphere — but the caller reaching this
// function has a cell held, and "no cell" is not a narrowing it should be
// asking about, so it is spelled out here as the whole ground rather than left
// to the library. An address that does not parse is not a place: nothing is
// standing in it, which is the conservative half of the disagreement — a map
// that draws nothing is wrong in a way a reader can see, where a server that
// narrows nothing while the seam narrows is wrong in a way nobody can.
func S2Held(mapping Mapping, token string) Held {
	if token == "" {
		return func(float64, float64) bool { return true }
	}
	cell, ok := s2Cell(token)
	if !ok {
		return func(float64, float64) bool { return false }
	}
	return func(x, y float64) bool {
		lat, lon := mapping.LatLng(x, y)
		return cell.Contains(s2.CellIDFromLatLng(s2.LatLngFromDegrees(lat, lon)))
	}
}

// S2CellAt is the leaf-most address of a point at the telescope's floor: the
// answer the pin card carries and `locate` in the port. It is empty when the
// point is nowhere the mapping can name.
func S2CellAt(mapping Mapping, x, y float64) string {
	lat, lon := mapping.LatLng(x, y)
	leaf := s2.CellIDFromLatLng(s2.LatLngFromDegrees(lat, lon))
	if !leaf.IsValid() {
		return ""
	}
	return leaf.Parent(deepestS2Level).ToToken()
}

// S2Level is the port level of an address: 0 for the root, 1 for a face, and
// so on down. A token that is not a place has no level, and says so with -1.
func S2Level(token string) int {
	if token == "" {
		return 0
	}
	cell, ok := s2Cell(token)
	if !ok {
		return -1
	}
	return cell.Level() + 1
}

// S2Parent is the address one level out. The root has no parent and neither
// does an address nobody can read; both answer the root, which is where a
// reader telescoping out of an address they cannot be standing in belongs.
//
// This is the arithmetic that makes S2 the proof the contract holds nothing
// geohash-shaped: an S2 address does not refine by appending characters, so
// telescoping out of `47a1cb` is not dropping its last character — it is
// `47a1`, three characters shorter and spelled differently.
func S2Parent(token string) string {
	cell, ok := s2Cell(token)
	if !ok || cell.Level() == 0 {
		return ""
	}
	return cell.Parent(cell.Level() - 1).ToToken()
}

// S2Children is the address's four children in their stable order, or the six
// faces when the address is the root. An address nobody can read has none.
func S2Children(token string) []string {
	if token == "" {
		out := make([]string, 0, 6)
		for face := range 6 {
			out = append(out, s2.CellIDFromFace(face).ToToken())
		}
		return out
	}
	cell, ok := s2Cell(token)
	if !ok {
		return nil
	}
	children := cell.Children()
	out := make([]string, 0, len(children))
	for _, child := range children {
		out = append(out, child.ToToken())
	}
	return out
}

// S2ParseCell reads an address the way the navigator's field does: the root is
// a place, a token deeper than the telescope's floor clamps to the floor, and
// anything else is either a canonically spelled token or not a place at all.
// It is the port's `parseInput` (analysis/cellsystems/s2.ts) minus the
// keystroke normalization, which belongs to the field and not to the server.
func S2ParseCell(text string) (string, bool) {
	if text == "" {
		return "", true
	}
	cell, ok := s2Cell(text)
	if !ok {
		return "", false
	}
	if cell.Level() > deepestS2Level {
		return cell.Parent(deepestS2Level).ToToken(), true
	}
	return cell.ToToken(), true
}

// s2Cell reads a token, refusing what is not a cell.
//
// The guard is not a formality. A token the library cannot read answers the
// zero cell id, whose descendant range is the whole of the space — so an
// unguarded `Contains` on it says yes to every point on the sphere, and a
// misspelled address would quietly stop narrowing anything.
func s2Cell(token string) (s2.CellID, bool) {
	cell := s2.CellIDFromToken(strings.TrimSpace(token))
	if cell == 0 || !cell.IsValid() {
		return 0, false
	}
	return cell, true
}
